jormungandr/src/notifier/mod.rs

```diff
@@ -0,0 +1,180 @@
+#[derive(Clone)]
+pub struct Notifier {
+    next_user_id: Arc<AtomicUsize>,
+    clients: Arc<tokio::sync::Mutex<Clients>>,
+    max_connections: usize,
+}
+
+#[derive(Serialize)]
+enum JsonMessage {
+    #[serde(serialize_with = "to_hex")]
+    NewBlock(HeaderId),
+    #[serde(serialize_with = "to_hex")]
+    NewTip(HeaderId),
+}
+
+type Clients = std::collections::HashMap<usize, warp::ws::WebSocket>;
+
+impl Notifier {
+    pub fn new(max_connections: Option<usize>) -> Notifier {
+        Notifier {
+            next_user_id: Arc::new(AtomicUsize::new(1)),
+            clients: Default::default(),
+            max_connections: max_connections.unwrap_or(MAX_CONNECTIONS_DEFAULT),
+        }
+    }
+
+    pub async fn start(&mut self, info: TokioServiceInfo, queue: MessageQueue<Message>) {
+        let clients1 = self.clients.clone();
+        let clients2 = self.clients.clone();
+        let logger = info.logger();
+
+        let (deleted_msgbox, deleted_queue) = channel::<usize>(32);
+
+        // TODO: it may be better to have a task that runs periodically instead of
+        // when a sender is detected to be disconected, but that would require
+        // reading the sockets besides from writing to them
+        info.spawn(
+            "clean disconnected notifier clients",
+            handle_disconnected(clients2.clone(), deleted_queue),
+        );
+
+        queue
+            .for_each(|input| {
+                info.spawn(
+                    "notifier send new messages",
+                    process_message(
+                        logger.clone(),
+                        clients1.clone(),
+                        input,
+                        deleted_msgbox.clone(),
+                    ),
+                );
+                futures::future::ready(())
+            })
+            .await;
+    }
+
+    pub async fn new_connection(&self, ws: warp::ws::WebSocket) {
+        let id = self.next_user_id.fetch_add(1, Ordering::Relaxed);
+
+        let clients = Arc::clone(&self.clients);
+
+        let rejected = async move {
+            let mut locked_clients = clients.lock().await;
+            if locked_clients.len() < (self.max_connections as usize) {
+                locked_clients.insert(id, ws);
+                None
+            } else {
+                Some(ws)
+            }
+        }
+        .await;
+
+        if let Some(mut ws) = rejected {
+            let close_msg = warp::ws::Message::close_with(
+                MAX_CONNECTIONS_ERROR_CLOSE_CODE,
+                MAX_CONNECTIONS_ERROR_REASON,
+            );
+            if ws.send(close_msg).await.is_ok() {
+                let _ = ws.close().await;
+            }
+        }
+    }
+}
+
+async fn process_message(
+    logger: Logger,
+    clients: Arc<tokio::sync::Mutex<Clients>>,
+    msg: Message,
+    mut disconnected: MessageBox<usize>,
+) {
+    let warp_msg = JsonMessage::from(msg).into();
+    let dead = async move { notify_all(clients, warp_msg).await };
```

Comment:
what is wrong with simply using:

```suggestion
    let dead = notify_all(clients, warp_msg).await;
```

Reply:
nothing, just a leftover of a little refactor